### PR TITLE
code examples for doc code snippets

### PR DESCRIPTION
**Pull request by 1sanz**
- code examples folder is added to docs
- this is where the code examples used in documentation will live

### CODE REVIEW

File: docs/codeexamples/writing_edsl.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
+# Copyright 2019 Intel Corporation.
```
Comment (flaub) [inline docs/codeexamples/writing_edsl.py]: Please update the copyright

File: docs/codeexamples/writing_edsl.cc [hunk 2]
```diff
@@ -0,0 +1,508 @@
+
+// Copyright 2019 Intel Corporation.
```
Comment (flaub) [inline docs/codeexamples/writing_edsl.cc]: Update copyright date

File: docs/codeexamples/writing_edsl.cc [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+
+!f32 = type tensor<!eltwise.f32>
+module {
+  func @sum_over_axis(%arg0: tensor<3x3x!eltwise.u64>) -> tensor<3x!eltwise.u64> {
+    %cst = "eltwise.sconst"() {value = 0.000000e+00 : f64} : () -> !f32
+    %0 = tile.cion add, none, %cst, %arg0 {sink = #map0, srcs = [#map1]} : !f32, tensor<3x3x!eltwise.u64> -> tensor<3x!eltwise.u64>
+    return %0 : tensor<3x!eltwise.u64>
+  }
+}
+)#"));
```
Comment (flaub) [inline docs/codeexamples/writing_edsl.cc]: Drop the `EXPECT_THAT` for all of these tests

File: docs/codeexamples/writing_edsl.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+!f32 = type tensor<!eltwise.f32>
+module {
+  func @sum_over_axis(%arg0: tensor<1x784x!eltwise.f32>) -> tensor<784x!eltwise.f32> {
+    %cst = "eltwise.sconst"() {value = 0.000000e+00 : f64} : () -> !f32
+    %0 = tile.cion add, none, %cst, %arg0 {sink = #map0, srcs = [#map1]} : !f32, tensor<1x784x!eltwise.f32> -> tensor<784x!eltwise.f32>
+    return %0 : tensor<784x!eltwise.f32>
+  }
+}
+'''
+        self.compare_results(program, expected)
```
Comment (flaub) [inline docs/codeexamples/writing_edsl.py]: Drop the `self.compare_results` for all for test tests. We only need to verify that it compiles at all, not that it generates this exact sequence of MLIR code.

File: docs/usage/writing_edsl.rst [hunk 5]
```diff
@@ -32,29 +33,14 @@ sum over axis `0` of a 2D tensor (in Keras this would be ``K.sum(I, axis=0)``):
...
-
-          def sum_over_axis(I):
-            M, N = TensorDims(2)
-            m, n = TensorIndexes(2)
-            I.bind_dims(M, N)
-            O = TensorOutput(N)
-            # contraction
-            O[n] += I[m, n]
-            return O
+      .. literalinclude:: ../codeexamples/writing_edsl.py
```
Comment (dgkutnic) [inline docs/usage/writing_edsl.rst]: Since we're including entire functions, can we use the built-ins that exist for the `literalinclude` directive? This will make it _much_ easier to maintain because we won't have to specify line numbers everywhere.

> The directive also supports including only parts of the file. If it is a Python module, you can select a class, function or method to include using the pyobject option:

```
.. literalinclude:: example.py
   :pyobject: Timer.start
```

[Read more...](https://www.sphinx-doc.org/en/1.5/markup/code.html#directive-literalinclude)

Comment (flaub) [reply docs/usage/writing_edsl.rst]: I think this only works for python, but i'm +100 on doing this for python :D

Comment (dgkutnic) [reply docs/usage/writing_edsl.rst]: It might not be much better than line numbers, but there are `start-after` and `end-before` directives for `literalinclude` that will at least let you use regexes to include lines of code rather than line numbers...

```
.. literalinclude:: ShowingCodeExamplesInSphinx.rst
   :language: rst
   :start-after: Instead of using
   :end-before: For example
```

[Read more...](https://build-me-the-docs-please.readthedocs.io/en/latest/Using_Sphinx/ShowingCodeExamplesInSphinx.html#literalinclude-directive)

**Comment by 1sanz** [comment]
@dgkutnic nice! :) done!

**Comment by 1sanz** [comment]
@flaub I am in the process of writing executable tests and cleaning out string style tests. Lets discuss more about upcoming documents and associated code file locations in person. Perhaps the code examples folder may need to move to plaidml/edsl/

**Comment by flaub** [comment]
Can we leave the montecarlo and other docs out of this PR and only have it deal with the EDSL for now? Then we can make a separate PR for the other docs.

File: docs/codeexamples/BUILD [hunk 6]
```diff
@@ -0,0 +1,56 @@
+load("@rules_python//python:defs.bzl", "py_binary")
```
Comment (flaub) [inline docs/codeexamples/BUILD]: Move all these rules to `plaidml/edsl/BUILD`

File: docs/BUILD [hunk 7]
```diff
@@ -17,6 +17,7 @@ py_binary(
     ],
     data = [
         ":runfiles",
+        "//docs/codeexamples:code_files",
```
Comment (flaub) [inline docs/BUILD]: I don't think we need this, let's have the `filegroup(name = 'docs')` defined in `plaidml/edsl/BUILD` include the `edsl_docs.cc` and `edsl_docs.py` files.

File: docs/codeexamples/BUILD [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        "monte_carlo_pi.py",
+        "monte_carlo_pi_tests.py",
+    ],
+    main = "monte_carlo_pi_tests.py",
+    deps = [
+        "//plaidml:py",
+        "//plaidml/edsl:py",
+        "//plaidml/exec:py",
+    ],
+)
```
Comment (flaub) [inline docs/codeexamples/BUILD]: Please remove the monte_carlo docs from this PR, we can do this in a separate PR.

File: docs/index.rst [hunk 9]
```diff
@@ -3,7 +3,7 @@
    You can adapt this file completely to your liking, but it should at least
    contain the root `toctree` directive.
 
-Welcome to plaidml's documentation!
+Plaidml Documentation
```
Comment (flaub) [inline docs/index.rst]: I sorta like the original text here. I also don't think we should ever use `Plaidml`, instead we should use `PlaidML`

File: docs/usage/writing_edsl.rst [hunk 10]
```diff
@@ -1,20 +1,21 @@
-Tile eDSL 
+Tile EDSL 
 #############
-The Tile eDSL (Embedded Domain Specific Language) provides developers with a
+The Tile EDSL (Embedded Domain Specific Language) provides developers with a
```
Comment (flaub) [inline docs/usage/writing_edsl.rst]: Can you leave it as `eDSL`?

File: docs/codeexamples/writing_edsl.cc [hunk 11]
```diff
@@ -0,0 +1,524 @@
+
```
Comment (flaub) [inline docs/codeexamples/writing_edsl.cc]: Please move this file to `plaidml/edsl/edsl_docs.cc`.

This file also doesn't compile (as shown by our buildkite build logs)

File: docs/codeexamples/writing_edsl.cc (hunk 3 above)
Comment (flaub) [reply docs/codeexamples/writing_edsl.cc]: I think we should drop all the asserts from this file.

File: docs/usage/edsl.rst [hunk 12]
```diff
@@ -403,13 +412,13 @@ or invalid set of index variables. For example, in the code:
 
   .. group-tab:: C++
 
-    .. literalinclude:: ../codeexamples/writing_edsl.cc
+    .. literalinclude:: ../../plaidml/edsl/edsl_doc.cc
       :language: cpp
       :lines: 133-136
```
Comment (flaub) [inline docs/usage/edsl.rst]: Still using `:lines:`, is that intended?

File: docs/usage/edsl.rst [hunk 13]
```diff
@@ -403,13 +412,13 @@ or invalid set of index variables. For example, in the code:
...
     
   .. group-tab:: Python
 
-    .. literalinclude:: ../codeexamples/writing_edsl.py
+    .. literalinclude:: ../../plaidml/edsl/edsl_doc.py
       :lines: 107-110
```
Comment (flaub) [inline docs/usage/edsl.rst]: Still using `:lines:`, is that intended?

File: docs/usage/edsl.rst [hunk 14]
```diff
@@ -597,14 +606,14 @@ directly to the formula, and so we get:
 
   .. group-tab:: C++
 
-    .. literalinclude:: ../codeexamples/writing_edsl.cc
+    .. literalinclude:: ../../plaidml/edsl/edsl_doc.cc
       :language: cpp
       :start-after: onv_2d_dilated_start
```
Comment (flaub) [inline docs/usage/edsl.rst]: is this a typo?

File: docs/usage/edsl.rst [hunk 15]
```diff
@@ -1,20 +1,21 @@
-Tile eDSL 
+Tile EDSL
```
Comment (flaub) [inline docs/usage/edsl.rst]: Can we use `eDSL` instead?

File: plaidml/edsl/edsl_doc.cc [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+#include "llvm/ADT/StringRef.h"
+
+using ::testing::ContainerEq;
+using ::testing::Eq;
+
+namespace plaidml::edsl {
+
+bool operator==(const Program& lhs, const std::string& rhs) {  //
+  return llvm::StringRef(lhs.str()).trim() == llvm::StringRef(rhs).trim();
+}
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: Drop this

File: plaidml/edsl/edsl_doc.cc [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+  // Specify the output size
+  auto O = TensorOutput(N, Y[0], Y[1], G, GCO);
+  // Compute the convolution
+  O(n, x[0], x[1], g, gco) +=
+      I(n, s[0] * x[0] + d[0] * k[0] - P[0], s[1] * x[1] + d[1] * k[1] - P[1], g, gci) * K(k[0], k[1], g, gci, gco);
+  return O;
+}
+// complex_conv_end
+
+}  // namespace plaidml::edsl
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: I think we should still have unit tests which ensure that all of this code actually compiles (by constructing a `Program`)

File: plaidml/edsl/edsl_doc.py [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+        P0 = ((Y0 - 1) * s0 + EK0 - X0) // 2
+        P1 = ((Y1 - 1) * s1 + EK1 - X1) // 2
+
+        # Specify the output size
+        O = TensorOutput(N, Y0, Y1, G, GCO)
+
+        # Compute the convolution
+        O[n, x0, x1, g, gco] += I[n, s0 * x1 + d0 * k0 - P0, s1 * x1 + d1 * k1 -
+                                  P1, g, gci] * K[k0, k1, g, gci, gco]
+        return O
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: We should still have unit tests that ensure things still compile (by constructing a `Program`)

File: docs/usage/edsl.rst (hunk 15 above)
Comment (dgkutnic) [reply docs/usage/edsl.rst]: A few members of the team expressed that `eDSL` was originally a typo and `EDSL` is the correct capitalization. That is why Sana capitalized them here. We should probably try and correct the understanding across the team if `eDSL` is the preferred capitalization.

I don't have a strong opinion either way, but can we at least be consistent across our documentation?

Comment (flaub) [reply docs/usage/edsl.rst]: Let's see if we can come to a consensus tomorrow when I'm in. In the meantime, I'd like to keep this particular PR clean from any changes in the copy (e.g. text).

Comment (dgkutnic) [reply docs/usage/edsl.rst]: SGTM

File: docs/usage/edsl.rst (hunk 14 above)
Comment (1sanz) [reply docs/usage/edsl.rst]: yes. thanks! fixed.

File: docs/usage/edsl.rst (hunk 13 above)
Comment (1sanz) [reply docs/usage/edsl.rst]: fixed.

File: docs/usage/edsl.rst (hunk 12 above)
Comment (1sanz) [reply docs/usage/edsl.rst]: fixed

File: plaidml/edsl/edsl_doc.py (hunk 18 above)
Comment (1sanz) [reply plaidml/edsl/edsl_doc.py]: done.

File: plaidml/edsl/edsl_doc.cc (hunk 17 above)
Comment (1sanz) [reply plaidml/edsl/edsl_doc.cc]: done.

File: plaidml/edsl/edsl_doc.cc (hunk 16 above)
Comment (1sanz) [reply plaidml/edsl/edsl_doc.cc]: done.

File: plaidml/edsl/edsl_doc.cc [hunk 19]
```diff
@@ -283,6 +281,22 @@ TEST(CppEdsl, AvgMerge) {
...
+TEST(CppEdsl, WrongMaxPool1D) {
+  auto I = Placeholder(DType::UINT64, {3, 3});
+  Program program("wrong_max_pool_1d", {WrongMaxPool1D(I)});
+}
+
+TEST(CppEdsl, ForLoopMaxPool1D) {
+  int N = 10;
+  float I[N], O[N / 2];
+  ForLoopMaxPool1D(I, O, N);
+}
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: We don't need a test for this, because this isn't EDSL code.

File: plaidml/edsl/edsl_doc.cc [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+  // Specify the output size
+  auto O = TensorOutput(N, Y[0], Y[1], G, GCO);
+  // Compute the convolution
+  O(n, x[0], x[1], g, gco) +=
+      I(n, s[0] * x[0] + d[0] * k[0] - P[0], s[1] * x[1] + d[1] * k[1] - P[1], g, gci) * K(k[0], k[1], g, gci, gco);
+  return O;
+}
+// complex_conv_end
+
+TEST(CppEdsl, SumOveAxis) {
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: Put these tests in some other fixture, we already have `CppEdsl` in another file.

File: plaidml/edsl/edsl_doc.cc [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+  auto I = Placeholder(DType::UINT64, {3, 3});
+  Program program("sum_over_axis", {SumOveAxis(I)});
+}
+
+TEST(CppEdsl, MaxOveAxis) {
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: type: MaxOverAxis

File: plaidml/edsl/edsl_doc.py [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+
+
+class TestEdslDocs(unittest.TestCase):
+    maxDiff = None
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: No need for this

File: plaidml/edsl/edsl_doc.py [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+import unittest
+
+import plaidml
+import plaidml.exec
+from plaidml.edsl import *
+
+
+class TestEdslHelper:
+
+    def sum_over_axis(I):
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: Remove all of these functions from the class

File: plaidml/edsl/edsl_doc.py [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+                                  P1, g, gci] * K[k0, k1, g, gci, gco]
+        return O
+
+
+class TestEdslDocs(unittest.TestCase):
+    maxDiff = None
+
+    def test_sum_over_axis(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 78]))
+        O = TestEdslHelper.sum_over_axis(I)
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: Call this like `O = sum_over_axis(I)`

File: plaidml/edsl/edsl_doc.cc [hunk 25]
```diff
@@ -0,0 +1,338 @@
+
+// Copyright 2020 Intel Corporation.
+// DO NOT TOUCH THIS FILE
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: We can drop this warning if it's no longer a requirement for line numbers to be exactly correct in the docs.

File: plaidml/edsl/edsl_doc.py [hunk 26]
```diff
@@ -0,0 +1,261 @@
+# Copyright 2020 Intel Corporation.
+# DO NOT TOUCH THIS FILE
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: Drop this warning, also update the comments to reflect the latest changes to the docs

File: plaidml/edsl/edsl_doc.cc [hunk 27]
```diff
@@ -0,0 +1,338 @@
...
+#include <gtest/gtest.h>
+
+#include "plaidml/edsl/edsl.h"
+#include "plaidml/exec/exec.h"
+#include "pmlc/util/env.h"
+#include "pmlc/util/logging.h"
+#include "llvm/ADT/StringRef.h"
+
+using ::testing::ContainerEq;
+using ::testing::Eq;
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: These are unused, drop them

File: plaidml/edsl/edsl_doc.cc [hunk 28]
```diff
@@ -0,0 +1,338 @@
+
+// Copyright 2020 Intel Corporation.
+// DO NOT TOUCH THIS FILE
+// Note: This file is being used by sphinx docs to pull in code blocks.
+//       Any changes made here may upset the docs.
+//       Code blocks are pulled into docs/usage/writing_edsl.rst if line numbers change here
+//       please update docs/usage/edsl.rst
+#include <float.h>
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: Drop this

File: plaidml/edsl/edsl_doc.cc [hunk 29]
```diff
@@ -0,0 +1,338 @@
...
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "plaidml/edsl/edsl.h"
+#include "plaidml/exec/exec.h"
+#include "pmlc/util/env.h"
+#include "pmlc/util/logging.h"
+#include "llvm/ADT/StringRef.h"
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: Drop these

File: plaidml/edsl/edsl_doc.py [hunk 30]
```diff
@@ -0,0 +1,261 @@
...
+#       please update docs/usage/edsl.rst
+
+import argparse
+import functools
+import os
+import sys
+import unittest
+
+import plaidml
+import plaidml.exec
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: Drop this

File: plaidml/edsl/edsl_doc.py [hunk 31]
```diff
@@ -0,0 +1,261 @@
...
+# Note: This file is being used by sphinx docs to pukl in code blocks.
+#       Any changes made here may upset the docs.
+#       Code blocks are pulled into docs/usage/writing_edsl.rst if line numbers change here
+#       please update docs/usage/edsl.rst
+
+import argparse
+import functools
+import os
+import sys
+import unittest
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: Drop unused imports

File: plaidml/edsl/edsl_doc.cc [hunk 32]
```diff
@@ -1,22 +1,15 @@
...
-// Note: This file is being used by sphinx docs to pull in code blocks.
-//       Any changes made here may upset the docs.
-//       Code blocks are pulled into docs/usage/writing_edsl.rst if line numbers change here
-//       please update docs/usage/edsl.rst
+// Note:
+//      This file is being used by sphinx docs to pull in code blocks.
+//      Code blocks are pulled into docs/usage/writing_edsl.rs
+//      Any changes made here may upset the docs.
+
 #include <float.h>
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.cc]: Please drop this

File: plaidml/edsl/edsl_doc.py [hunk 33]
```diff
@@ -1,261 +1,268 @@
...
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[()] += I[x, y]
+    return Sum / (X * Y)
+
+
+def for_loop_max_pool_1d():
+    # for_loop_max_pool_1d_start
+    for i in range(1, N // 2):
+        curr_max = numpy.finfo(float).eps
```
Comment (flaub) [inline plaidml/edsl/edsl_doc.py]: Is there a way we can avoid using numpy for this example?

File: plaidml/edsl/edsl_doc.cc (hunk 32 above)
Comment (flaub) [reply plaidml/edsl/edsl_doc.cc]: Please drop the `<float.h>` but leave the comments
